uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/instances/leave-request-instance_test.js

```diff
@@ -1,137 +1,284 @@
+          expect(error).toEqual(jasmine.objectContaining(expectedError));
         });
-      })
+      });
+    });
+  });
+
+  describe('create()', function () {
+    var instanceCreate;
+
+    beforeEach(function () {
+      requestData = helper.createRandomLeaveRequest();
+      instance = LeaveRequestInstance.init(requestData, true);
+      instanceCreate = instance.create();
+    });
+
+    afterEach(function () {
+      //to excute the promise force an digest
+      $rootScope.$apply();
     });
 
-    describe('update()', function () {
-      var attr,
-        sendPOSTMock,
-        updatePromise;
+    it('calls equivalent API method', function () {
+      instanceCreate.then(function () {
+        expect(LeaveRequestAPI.create).toHaveBeenCalled();
+      });
+    });
+
+    it('id is appended to instance', function () {
+      expect(instance.id).not.toBeDefined();
+      instanceCreate.then(function () {
+        expect(instance.id).toBeDefined();
+        expect(instance.id).toEqual(jasmine.any(String));
+      });
+    });
+
+    describe('when one mandatory filed is missing', function () {
 
       beforeEach(function () {
-        attr = {
-          key: jasmine.any(String)
+        expectedError = {
+          is_error: 1,
+          error_message: 'contact_id, from_date and from_date_type in params are mandatory'
         };
-        sendPOSTMock = jasmine.createSpy('sendPOST');
-        sendPOSTMock.and.returnValue(jasmine.any(Promise));
-        LeaveRequestAPI.sendPOST = sendPOSTMock;
+        requestData = helper.createRandomLeaveRequest();
+        delete requestData.contact_id;
+        _.assign(instance, requestData);
+        instanceCreate = instance.create();
+      });
+
+      afterEach(function () {
+        //to excute the promise force an digest
+        $rootScope.$apply();
+      });
+
+      it('fails to create instance', function () {
+        instanceCreate.catch(function (error) {
+          expect(error).toEqual(jasmine.objectContaining(expectedError));
+        });
+      });
+    });
+  });
+
+  describe('isValid()', function () {
+    var instanceValid;
+
+    beforeEach(function () {
+      requestData = {
+        contact_id: 123
+      };
+      instance = LeaveRequestInstance.init(requestData, true);
+      instanceValid = instance.isValid();
+    });
 
-        updatePromise = LeaveRequestInstance.update(attr);
+    afterEach(function () {
+      //to excute the promise force an digest
+      $rootScope.$apply();
+    });
+
+    it('calls equivalent API method', function () {
+      instanceValid.then(function () {
+        expect(LeaveRequestAPI.isValid).toHaveBeenCalled();
       });
+    });
 
-      it('calls LeaveRequestAPI.sendPOST', function () {
-        expect(LeaveRequestAPI.sendPOST).toHaveBeenCalledWith('LeaveRequest', 'create', _.assign({}, LeaveRequestInstance.attributes(), attr));
+    describe('when leave request is valid', function () {
```

Comment:
testing the various responses which will be sent by the server for isValid calls